### ISSUE TITLE
prepareNonServerSharedMemory error

### ISSUE THREAD

**Issue by gmuscianisi** [irods/irods#4675]
Dear all, 
in my rodslog I'm having the following error

ERROR: prepareNonServerSharedMemory: failed to get shared memory object [irods_re_cache_shared_memory_upgraded_irods_rule_language_rule_engine_pid21941_00a5f8c85bbe808318cd74a9a5bf9b4e7d5334bab5bb5a20b76e52af3ac64bd39f6e9820f1fb6cc8]. Exception caught [No such file or directory]

Could you give me some suggestions how to solve it? What can I check in my server configuration to solve it, or other?
My irods server version is 4.2.0

Thanks in advance. Giusy

**Comment by trel** [comment]
Is this on a CentOS7.3 system?  `/dev/shm` might have the wrong permissions...

https://bugzilla.redhat.com/show_bug.cgi?id=1406254

**Comment by trel** [comment]
Also, consider running 4.2.7 - we've fixed a great many things.

**Comment by kript** [comment]
We are seeing this consistently across our zones too, all Ubuntu 18.04, 4.2.7.
Some zones with quotas, most without. Any further triage or logs I can help with?

**Comment by kript** [comment]
I should add that while it is spread over all the zones, it is not consistent with activity or a particular rule that I can see, from a very basic analysis. Its also intermittent - on one zone we have it occurring a lot, then nothing for over a week.

**Comment by trel** [comment]
Not sure how to address this unless we can reproduce it.

No known bugs in the current implementation as far as I know.

**Comment by tedgin** [comment]
We started seeing this today running iRODS 4.1.10 on CentOS 7.9. It's only happening on one of our test systems so far.

```
Apr 22 14:04:30 pid:2207 NOTICE: Agent process 2378 started for puser=rods and cuser=rods from 128.196.65.86
Apr 22 14:04:30 pid:2378 ERROR: prepareNonServerSharedMemory: failed to get shared memory object [irods_re_cache_shared_memory_pid2207_03647a6a80a4c3a798d6dbcf7eacfe4836093b940e1c179b0ea659ae42b0a6fd36943dc56faa01cb]. Exception caught [No such file or directory]
Apr 22 14:04:31 pid:2378 NOTICE: readAndProcClientMsg: received disconnect msg from client
Apr 22 14:04:31 pid:2378 NOTICE: Agent exiting with status = 0
Apr 22 14:04:31 pid:2207 NOTICE: Agent process 2378 exited with status 0
```

**Comment by tedgin** [comment]
I was able to get rid of them by disabling the irods service in systemd, .i.e., `systemctl disable irods`.  After that, I rebooted the server and manually started iRODS using `irodsctl`.

**Comment by trel** [comment]
Yes, any leftover shared memory object files are cleaned on server startup.

Still not sure how/when this is happening in the first place.

**Comment by tedgin** [comment]
The key point of my fix was keeping systemd from starting iRODS.  Before I did that, restarting iRODS and rebooting the server didn't stop the error from occuring.

I started seeing this problem after I updated the packages installed on the iRODS server.  I have a hunch that something in systemd was updated that broke iRODS's access to shared memory.

**Comment by trel** [comment]
Oooh, I see.    Thanks.

**Comment by kript** [comment]
Actually, I have been able to reproduce something very similar using the `irods_demo` repo see my [PR](https://github.com/irods/irods_demo/pull/23), might be of some assistance in reproducing?

It ddn't have anything to do with `systemd`, but that doesn't mean systemd isn't doing its own thing to 'help'...

**Comment by gmuscianisi** [comment]
Dear all, 
I have followed the suggestion described in the thread, I mean: 
-- I have checked the permissions of /dev/shm and these are 777 
-- I have disabled the irods service in systemd, restarted the server and then manually started iRODS using irodsctl 

But I obtain the same  irods error in the rodslog, pasted below:

Jun 23 12:30:05 pid:8224 remote addresses: 130.186.17.95 ERROR: prepareNonServerSharedMemory: failed to get shared memory object [irods_re_cache_shared_memory_upgraded_irods_rule_language_rule_engine_pid21809_78aa66f98b84196e46c51a9e05555fcddf3f6df98ce49f8d3273c38159806fb1b5f9d25308f0d055]. Exception caught [No such file or directory]
 

What else shoud I check or modify to solve this problem? 

The OS of the server is CentOS Linux release 7.2.1511 (Core) and the iRODS server is 4.2.0 

Thanks a lot. Giusy

**Comment by trel** [comment]
If you stop the iRODS server - please remove any/all `/dev/shm/irods_*` files, they shouldn't be there if the server is not running.

Also, consider upgrading to 4.2.8.       We are working to get 4.2.10 out ASAP, there is one known regression in 4.2.9.

**Comment by gmuscianisi** [comment]
Dear Terrell, 
thanks for your reply. The /dev/shm directory is always empty on my irods server both when the server is running or stopped, so I don't have any file to remove. 
Where can I check for the files you have cited? 

We are planning to upgrade, but we need time for this operation.
Thanks. Giusy

**Comment by trel** [comment]
That suggests the permissions are still not correct on that `/dev/shm` directory.   The irods service account should be creating and managing its shared memory in that location.

**Comment by gmuscianisi** [comment]
The permission of the location is configured as below:
 
[idors-server:~]$ ls -lrtd /dev/shm/
drwxrwxrwt 2 root root 40 Jun 23 17:20 /dev/shm/

Is there anything else I can check? 

Thanks. Giusy

**Comment by tedgin** [comment]
This feels like an selinux issue. Is it possible that selinux is preventing iRODS from allocating shared memory?

**Comment by gmuscianisi** [comment]
Dear tedgin, thanks for your reply. 
I have checked and selinux is disabled on my irods server. 
Moreover, the irods admin user is able to write on /dev/shm (I tried writing something in this area with the irods admin user). 

Any other idea?  Thanks a lot. Giusy

**Comment by kript** [comment]
We only seem to see this on one production system, and that's the one with Quota and active use of them. Is that the same for you @tedgin @gmuscianisi ?

i.e. does the Provider showing this issue have `acRescQuotaPolicy {msiSetRescQuotaPolicy("on"); }` set?

**Comment by tedgin** [comment]
I have only seen this in one of our development environments.

**Comment by trel** [comment]
Is this just happening when there's not enough available RAM on the machine?

Feels like a resource starvation issue if it's not permissions.

**Comment by kript** [comment]
IIRC I checked both at the time when I caught it and over time on our resource monitoring system; there was enough RAM available. However, shared memory allocation is not the same as free memory, and we may need to allocate more (as we did for the Docker PR above).

**Comment by kript** [comment]
> We only seem to see this on one production system, and that's the one with Quota and active use of them. Is that the same for you @tedgin @gmuscianisi ?
> 
> i.e. does the Provider showing this issue have `acRescQuotaPolicy {msiSetRescQuotaPolicy("on"); }` set?

We have _also_ seen this on systems that have run out of memory, FWIW.

**Comment by alanking** [comment]
For anybody who runs across this...

In a recent support call with a user experiencing this, we observed that systemd was repeatedly deleting the iRODS shared memory files which explains why so many agents were getting stuck. It was discovered that the iRODS service account user was not a system user, which means that systemd will clean up shared memory files owned by that user. It's possible that the system got this way because the iRODS service account user was created before the setup script created the user as a system user (the setup script invokes `useradd -r`, where `-r` indicates that it is a system user).

We at least temporarily resolved the issue in this case by modifying `/etc/systemd/logind.conf` to explicitly set `RemoveIPC=no`.

**Comment by trel** [comment]
I think the issue, not seen since 4.2.7, can be safely assumed to now be handled better in later versions.

The process model refactor in 5.0.0 gives me even more confidence that this is not something we need to take action on.

Closing.
